test/src/dftbp/api/mm/testers/test_expdmhs.c

```diff
@@ -0,0 +1,168 @@
+  
+  int natom = dftbp_get_nr_atoms(&calculator);
+  fprintf(atf, "natom       :integer:0:\n%d\n",natom);
+  
+  dftbp_set_coords_and_lattice_vecs(&calculator, coords_si2, latvecs_si2);
+  
+  int basis_size = dftbp_get_basis_size(&calculator);
+  fprintf(atf, "basis_size       :integer:0:\n%d\n", basis_size);
+  
+  _Bool is_hs_real = dftbp_is_hs_real(&calculator);
+  fprintf(atf, "is_hs_real       :integer:0:\n%d\n", (is_hs_real ? 1 : 0) );
+  
+  int n_spin = dftbp_get_nr_spin(&calculator);
+  fprintf(atf, "n_spin       :integer:0:\n%d\n", n_spin);
+  
+  int n_kpts = dftbp_nr_kpoints(&calculator);
+  fprintf(atf, "n_kpts       :integer:0:\n%d\n", n_kpts);
+  
+  double kweights[N_KPTS];
+  dftbp_get_kweights(&calculator, kweights);
+  fprintf(atf, "kweights       :real:1:4\n");
+  fprintf(atf, "%f %f %f %f\n", kweights[0], kweights[1], kweights[2], kweights[3]);
+  
+  int nr_local_ks = dftbp_get_nr_local_ks(&calculator);
+  fprintf(atf, "nr_local_ks       :integer:0:\n%d\n", nr_local_ks); 
+
+  int local_ks[nr_local_ks * 2];
+  dftbp_get_local_ks(&calculator, local_ks);
+  fprintf(atf, "local_ks       :integer:2:%d,%d\n", nr_local_ks, 2);
+  for (int i = 0; i < nr_local_ks; ++i)
+  {
+    fprintf(atf, "%i %i\n", local_ks[i * 2], local_ks[i * 2 + 1]);
+  }
+
+  dftbp_register_dm_callback(&calculator, dm_callback, 0);
+  dftbp_register_s_callback(&calculator, s_callback, 0);
+  dftbp_register_h_callback(&calculator, h_callback, 0);
+  
+  assert(BASIS_SIZE == basis_size);
+  assert(N_KPTS == n_kpts);
+  assert(N_SPIN == n_spin);
+
+  /* Evaluate energy */
+  double mermin_energy;
+  dftbp_get_energy(&calculator, &mermin_energy);
+  fprintf(atf, "mermin_energy       :real:0:\n%f\n",mermin_energy);
+  
+  
+  for (int iK = 1; iK <= N_KPTS; ++iK)
+  for (int iS = 1; iS <= N_SPIN; ++iS)
+  {
+    fprintf(atf, "dm_%d_%d       :complex:2:%d,%d\n", iK, iS, BASIS_SIZE, BASIS_SIZE);
+    print_matrix(atf, &(dm[iK-1][iS-1][0][0]));
+
+    fprintf(atf, "overlap_%d_%d       :complex:2:%d,%d\n", iK, iS, BASIS_SIZE, BASIS_SIZE);
+    print_matrix(atf, &(overlap[iK-1][iS-1][0][0]));
+
+    fprintf(atf, "hamiltonian_%d_%d       :complex:2:%d,%d\n", iK, iS, BASIS_SIZE, BASIS_SIZE);
+    print_matrix(atf, &(hamiltonian[iK-1][iS-1][0][0]));
+    
+    double complex Hsum=0, Ssum=0;
+    for (int i = 0; i < BASIS_SIZE; ++i)
+    {
+      for (int j = 0; j < BASIS_SIZE; ++j)
+      {
+        double complex d = (i < j) ? dm[iK-1][iS-1][i][j] : conj(dm[iK-1][iS-1][j][i]);
+        double complex s  = (i < j) ? overlap[iK-1][iS-1][i][j] : conj(overlap[iK-1][iS-1][j][i]);
+        double complex h  = (i < j) ? hamiltonian[iK-1][iS-1][i][j] : conj(hamiltonian[iK-1][iS-1][j][i]);
+        Ssum += d * s;
+        Hsum += d * h;
+      }
+    }
+    fprintf(atf, "Ssum_%d_%d       :complex:0:\n%f %f\n", iK, iS, creal(Ssum), cimag(Ssum));
+    fprintf(atf, "Hsum_%d_%d       :complex:0:\n%f %f\n", iK, iS, creal(Hsum), cimag(Hsum));
+  }
+
+  fclose(atf);
+  return 0;
+}
```

Comment:
The problem with all tests is, that they won't fail, if DFTB+ passed the wrong data!

Reply:
I don't understand what you mean. _autotest.tag contains complete DM, H, and S matrices. If anything changes, test will fail. The only problem that I would fix is summations  along k-points (Ssum and Hsum) - currently it is physically meaningless, but still works as a checksum.

Reply:
It will take a time to implement proper summations over k-points. Also design decision should be made about k-grid import/export (what should it be? both?).  So I close this conversation as resolved and will address that issue in a follow-up pull request.